### PR TITLE
Implement vars.json expansion in test data

### PR DESCRIPTION
**Pull request by jknphy**
See [poo#88059](https://progress.opensuse.org/issues/88059)

### CODE REVIEW

File: t/02_yaml_schedule.t [hunk 1]
```diff
@@ -21,6 +21,17 @@ subtest 'parse_yaml_test_data_single_import' => sub {
...
 
+subtest 'parse_yaml_test_data_vars_expansion' => sub {
+    use scheduler;
+    use testapi 'set_var';
+
+    set_var('ENV_VAR', 'my-env-details');
+    my $schedule = $ypp->load_file(dirname(__FILE__) . '/data/test_data_vars_expansion.yaml');
+    scheduler::parse_test_suite_data($schedule);
+    my $testdata = scheduler::get_test_suite_data();
+    ok $testdata->{name_with_env_details} eq 'name-my-env-details', "Test data value was expanded properly";
```
Comment (rwx788) [inline t/02_yaml_schedule.t:32]: I believe we should cover more cases than just one. At least some case with nested key, as well as array.

File: lib/scheduler.pm [hunk 2]
```diff
@@ -86,6 +86,24 @@ sub get_test_suite_data {
...
+
+Expand test suite data variables
+
+=cut
+
+sub expand_test_data_vars {
+    my ($node) = shift;
+    if (ref $node eq 'HASH') {
+        foreach my $k (keys %$node) {
+            $node->{$k} =~ s/%(.*?)%/get_var($1)/eg if !ref $node->{$k};
```
Comment (rwx788) [inline lib/scheduler.pm:99]: What's about the case if variable is not defined? Did you check what openQA does in such case? In anycase we should not fail here with attempt to replace line with undef. I would propose either set empty string as default value for `get_var`, or not to define it, depending on what openQA does to be consistent.

File: test_data/yast/skip_registration/default_repos.yaml [hunk 3]
```diff
@@ -15,10 +15,8 @@ repos:
     uri: cd:/\?devices=/dev/disk/by-id/scsi-0QEMU_QEMU_CD-ROM_cd0
     enabled: No
     filter: name
-  - name: SLES
-    nr: 3
+  - name: 'SLES-%VERSION%'
```
Comment (rwx788) [inline test_data/yast/skip_registration/default_repos.yaml:18]: Please, split this change to the separate commit.

File: lib/scheduler.pm (hunk 2 above)
Comment (jknphy) [reply lib/scheduler.pm:99]: I checked, it didn't fail, for example it printed: "SLES-" without expansion.

Comment (rwx788) [reply lib/scheduler.pm:99]: Do you have logs, because I've written small code snipper:
```perl
sub mmm {
    return undef;
}

my $ttt = "%AAA%";
$ttt =~ s/%(.*?)%/mmm()/eg;
```
And I get `Use of uninitialized value in substitution iterator`. Documentation and the code of os-autoinst will return `undef`. I guess you mean that program didn't crash, but I'm quite confident that you had this exception in the output somewhere, like we have for concat of undef sometimes.

Comment (rwx788) [reply lib/scheduler.pm:99]: And yeah, following code:
```perl
sub mmm {
    return undef;
}

my $ttt = "bzz%AAA%";
$ttt =~ s/%(.*?)%/mmm()/eg;

print($ttt);
```
Prints `bzz`, but still with the error being thrown.

Comment (jknphy) [reply lib/scheduler.pm:99]: I can see it in the logs, good catch. let me check.

File: test_data/yast/skip_registration/default_repos.yaml (hunk 3 above)
Comment (jknphy) [reply test_data/yast/skip_registration/default_repos.yaml]: doesn't look useful for this use case, after version we need [15.3-0](https://openqa.suse.de/tests/5421445#step/validate_repos/12) which is not in vars.json. I dropped the commit for now.

File: t/02_yaml_schedule.t (hunk 1 above)
Comment (jknphy) [reply t/02_yaml_schedule.t]: done, good suggestion, thank you, I discovered some bug in my initial design.

File: lib/scheduler.pm [hunk 4]
```diff
@@ -86,6 +86,24 @@ sub get_test_suite_data {
...
+=cut
+
+sub expand_test_data_vars {
+    my ($node) = shift;
+    if (ref $node eq 'HASH') {
+        $_ = expand_test_data_vars($_) foreach values %$node;
+    } elsif (ref $node eq 'ARRAY') {
+        $_ = expand_test_data_vars($_) foreach (@$node);
+    } else {
+        $node =~ s/%(.*?)%/get_var($1,'')/eg;
```
Comment (rwx788) [inline lib/scheduler.pm:102]: Can you confirm that it's what happens when we have `%VAR%` in vars.json? I would also suggest adding test case for the case when variable is not defined.

Comment (jknphy) [reply lib/scheduler.pm:102]: in vars.json the values are already replaced. I tried to use '%VAR%' in yaml in openQA and it replaced with empty string as expected because it doesn't exist. Is what you meant?
I added test cases for empty value.

Comment (rwx788) [reply lib/scheduler.pm:102]: Yep, wanted to confirm that we implement same behavior as openQA, so expectations are clear. Thanks!

**Review by rwx788 — APPROVED** [review]
LGTM
